File: templates/Py/trie.py
class TrieNode:
  def __init__(self):
    self.is_word = False
    self.children = dict()


class Trie:
  def __init__(self):
    self.root = TrieNode()

  def insert(self, word):
    cur = self.root
    for c in word:
      if c in cur.children:
        cur = cur.children[c]
      else:
        new_node = TrieNode()
        cur.children[c] = new_node
        cur = new_node
    cur.is_word = True

  def search(self, word) -> bool:
    cur = self.root
    for c in word:
      if c not in cur.children:
        return False
      else:
        cur = cur.children[c]
    return cur.is_word

  def startsWith(self, prefix) -> bool:
    cur = self.root
    for c in prefix:
      if c not in cur.children:
        return False
      else:
        cur = cur.children[c]
    return True

File: templates/C++/trie.cpp
#include <string>
#include <unordered_map>

using namespace std;

struct TrieNode {
  std::unordered_map<char, TrieNode *> children{};
  bool is_word = false;
};

class Trie {
private:
  TrieNode *root;

public:
  Trie() { root = new TrieNode(); }

  void insert(string word) {
    auto p = root;
    for (const auto &c : word) {
      if (p->children.count(c) == 0)
        p->children[c] = new TrieNode();
      p = p->children[c];
    }
    p->is_word = true;
  }

  bool search(string word) {
    auto p = root;
    for (const auto &c : word) {
      if (p->children.count(c) == 0)
        return false;
      p = p->children[c];
    }
    return p->is_word;
  }

  bool startsWith(string prefix) {
    auto p = root;
    for (const auto &c : prefix) {
      if (p->children.count(c) == 0)
        return false;
      p = p->children[c];
    }
    return true;
  }
};
